app/utilities/drawing/utils.py

```diff
@@ -0,0 +1,125 @@
+import sys
+import PIL
+from PIL import Image, ImageDraw, ImageOps, ImageFont, ImageEnhance, ImageFilter
+
+from . import constants
+
+
+def drawBadgeBackgroundOnBackground(offset, background):
+    black_region = PIL.Image.new(mode="RGBA", size=constants.badge_size, color=(255, 255, 255, 25))
+    background.paste(black_region, (offset[0], offset[1]), create_rounded_rectangle_mask(black_region, 8))
+    return background
+
+def drawBadgeImageOnBackground(offset, background, badge_name):
+    badge_path = constants.badges_map[badge_name]
+    black_region = Image.open(badge_path)
+    black_region = black_region.resize(constants.badge_size, Image.Resampling.LANCZOS)
+    background.paste(black_region, (offset[0], offset[1]), create_rounded_rectangle_mask(black_region, 8))
+    return background
+
+def createSemiTransparentBlackRegionOnBackground(size, offset, background):
+    cropped_img = background.crop((
+        offset[0],
+        offset[1],
+        size[0]+offset[0],
+        size[1]+offset[1]
+    ))
+    alpha_bg = PIL.Image.new(mode="RGBA", size=cropped_img.size, color=(18, 17, 21, 127))
+    black_region = Image.alpha_composite(cropped_img, alpha_bg)
+    background.paste(black_region, (offset[0], offset[1]), create_rounded_rectangle_mask(black_region, 12))
+    return background
+
+def applyAlphaWithMask(image, mask_path):
+    mask = Image.open(mask_path).convert('L')
+    output = ImageOps.fit(image, mask.size, centering=(0.5, 0.5))
+    output.putalpha(mask)
+    return output
+
+def drawText(text, size, location, weight, draw, color=(255,255,255)):
+    font_path = constants.default_font
+    font = ImageFont.truetype(font_path, size)
+    font.set_variation_by_name(weight)
+    draw.text(location, text, color, font=font)
+
+def create_rounded_rectangle_mask(rectangle, radius):
+    solid_fill =  (50,50,50,255)
+    # create mask image. all pixels set to translucent
+    i = Image.new("RGBA",rectangle.size,(0,0,0,0))
+
+    # create corner
+    corner = Image.new('RGBA', (radius, radius), (0, 0, 0, 0))
+    draw = ImageDraw.Draw(corner)
+    # added the fill = .. you only drew a line, no fill
+    draw.pieslice((0, 0, radius * 2, radius * 2), 180, 270, fill = solid_fill)
+
+    # max_x, max_y
+    mx,my = rectangle.size
+
+    # paste corner rotated as needed
+    # use corners alpha channel as mask
+
+    i.paste(corner, (0, 0), corner)
+    i.paste(corner.rotate(90), (0, my - radius),corner.rotate(90))
+    i.paste(corner.rotate(180), (mx - radius,   my - radius),corner.rotate(180))
+    i.paste(corner.rotate(270), (mx - radius, 0),corner.rotate(270))
+
+    # draw both inner rects
+    draw = ImageDraw.Draw(i)
+    draw.rectangle( [(radius,0),(mx-radius,my)],fill=solid_fill)
+    draw.rectangle( [(0,radius),(mx,my-radius)],fill=solid_fill)
+
+    return i
+
+def ReduceOpacity(im, opacity):
+    assert opacity >= 0 and opacity <= 1
```

Comment:
Either clamp opacity to [0, 1] or modulo it within the range. We shouldn't be asserting for this. I'd probably say clamping it since that's how most programs deal with it.